[Python] Update docs for operators and streams (#204)
erdos-project/erdos: python/doc/source/operators.rst

Operators
=========

> ERDOS operators process received data, and use streams to broadcast Messages 

Reviewer: Could you link to the `Message` and `WatermarkMessage` classes?

to downstream operators.
We provide a standard library of operators for common dataflow patterns
under :py:mod:`erdos.operator`.
While the standard operators are general and versatile, some applications may
implement custom operators to better optimize performance and take
fine-grained control over exection.

> Operators are structures which implement a certain communication pattern. For example, 

Reviewer: Could you add that operators are implemented as classes, and must subclass the built-in operator type that specifies its communication pattern?

Also, describe the different types of communication patterns (Source, Sink, OneInOneOut, etc) in an unordered list

the `SendOp` from 
`python/examples/simple_pipeline.py <https://github.com/erdos-project/erdos/blob/master/python/examples/simple_pipeline.py>`_ 
implements a :py:class:`erdos.operator.Source` operator because it does not receive any data, 
and sends messages on a single output stream.

Operators can support both push and pull-based models of execution by 
implementing methods defined for each operator. By implementing callbacks 
such as :py:meth:`erdos.operator.OneInOneOut.on_data()`, operators can process 
messages as they arrive. Moreover, operators can implement callbacks over watermarks 
(e.g. :py:meth:`erdos.operator.OneInOneOut.on_watermark()`) to ensure ordered 
processing over timestamps. ERDOS ensures lock-free, safe, and concurrent processing 

> by ordering callbacks in an ERDOS-managed execution lattice, which serves as a 

Reviewer: ```suggestion
via a system-managed ordering of callbacks, which is implemented as a
```
